orm/src/main/java/org/hibernate/search/v6poc/entity/orm/model/impl/HibernateOrmPathFilterFactory.java

```diff
@@ -0,0 +1,290 @@
+					return Optional.empty();
+				}
+				else {
+					return Optional.of( baseValue );
+				}
+			}
+		}
+		else if ( SimpleValue.class.equals( valueClass ) ) { // equals() and not isAssignableFrom(), we mean it.
+			pathsAsStrings.add( propertyNode.toPropertyString() );
+			// The string representation of the path was added, we can stop here
+			return Optional.empty();
+		}
+		else if ( SimpleValue.class.isAssignableFrom( valueClass ) ) {
+			if ( isWholePath && ToOne.class.isAssignableFrom( valueClass )
+					|| isWholePath && Any.class.isAssignableFrom( valueClass ) ) {
+				// The path as a whole (and not just a prefix) was resolved to an association
+				pathsAsStrings.add( path.getParent().toPropertyString() );
+				// The string representation of the path was added, we can stop here
+				return Optional.empty();
+			}
+			else {
+				return Optional.of( baseValue );
+			}
+		}
+		else if ( org.hibernate.mapping.Collection.class.isAssignableFrom( valueClass ) ) {
+			if ( extractorPath.isEmpty() ) {
+				throw log.unknownPathForDirtyChecking( persistentClass.getMappedClass(), propertyNode, null );
+			}
+
+			List<? extends Class<? extends ContainerValueExtractor>> extractorClasses =
+					extractorPath.getExplicitExtractorClasses();
+			Iterator<? extends Class<? extends ContainerValueExtractor>> extractorClassIterator =
+					extractorClasses.iterator();
+
+			Value containedValue = baseValue;
+			org.hibernate.mapping.Collection collectionValue;
+			do {
+				collectionValue = (org.hibernate.mapping.Collection) containedValue;
+				try {
+					containedValue = resolveContainedValue( collectionValue, extractorClassIterator.next() );
+				}
+				catch (SearchException e) {
+					throw log.unknownPathForDirtyChecking( persistentClass.getMappedClass(), path, e );
+				}
+			}
+			while ( extractorClassIterator.hasNext() && containedValue instanceof org.hibernate.mapping.Collection );
+
+			if ( !extractorClassIterator.hasNext() ) {
+				// We managed to resolve the whole container value extractor list
+				Class<? extends Value> containedValueClass = containedValue.getClass();
+				if ( SimpleValue.class.equals( containedValueClass ) // equals() and not isAssignableFrom(), we mean it.
+						|| Component.class.isAssignableFrom( containedValueClass )
+						|| isWholePath && OneToMany.class.isAssignableFrom( containedValueClass )
+						|| isWholePath && ToOne.class.isAssignableFrom( containedValueClass )
+						|| isWholePath && Any.class.isAssignableFrom( containedValueClass ) ) {
+					pathsAsStrings.add( propertyNode.toPropertyString() );
+					pathsAsStrings.add( collectionValue.getRole() );
+					// The string representation of the path was added, we can stop here
+					return Optional.empty();
+				}
+				else {
+					return Optional.of( containedValue );
+				}
+			}
+		}
+
+		throw log.unknownPathForDirtyChecking( persistentClass.getMappedClass(), path, null );
+	}
```

Comment:
Seeing this method, I'm wondering if all this complicated logic should be in ORM instead of Search?

Reply:
ORM doesn't really have a unified concept of "dirty path": sometimes it sends us a `PostUpdateEvent` with a list of dirty properties, sometimes an `AbstractCollectionEvent` with a String representing the role of the collection, ...

This code pretty much complements our `FullTextIndexEventListener`, guessing what kind of events we might receive and how the paths will be represented in each event. So I'm not sure it would make sense to push this to ORM without also pushing the listener, which sort of "unifies" events to only consider dirty paths. And I doubt that's something the ORM team will want: the event system is quite complicated, there must be a reason.

Maybe we could move *part* of the logic, such as adding `isAssociation()` getters. But walking the property tree would have to stay in Search, I think, mainly because our representation of paths is more powerful than the one in ORM: I'm not sure you can have an association wrapped in an `Optional` in ORM, for instance.

I think ORM 6 introduces a new way of visiting the metamodel, using visitors that are more advanced than what we have today. Not sure it will make the code clearer, but we can give it a try.